File: activity/20200904/RabbitLee/Best_Time_to_Buy_and_Sell_Stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        if (n <= 1) return 0;
        
        int res = 0;
        int low = prices[0];
        for (int i = 1; i < n; i++) {
            res = max(res, prices[i] - low);
            low = min(low, prices[i]);
        }
        return res;
    }
};

File: activity/20200904/scau-flyingbird/BestTimeToBuyAndSellStock.go
func maxProfit(prices []int) int {
	if len(prices) <= 1 {
		return 0
	}
	min, max := prices[0], 0
	for i := 1; i < len(prices); i++ {
		max = Max(max, prices[i]-min)
		min = Min(min, prices[i])
	}
	return max
}

func Min(a, b int) int {
	if a < b {
		return a
	} else {
		return b
	}
}

func Max(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}
